### PR TITLE
enable attributes overriding

### PR DESCRIPTION
**Pull request by yangcao77**
Signed-off-by: Stephanie <yangcao@redhat.com>

### What does this PR do?
This PR enables attributes overriding in commands,components and projects. 
Adds attributes fields in override objects.

Also change back to allow additional properties in metadata 

### What issues does this PR fix or reference?
#233 
#236

### CODE REVIEW

**Review by maysunfaisal — COMMENTED** [review]
lgtm

**Comment by yangcao77** [comment]
@davidfestal @amisevsk May I get a 'lgtm' for this PR? Maysun's approval did not pass the review requirement.

**Review by maysunfaisal — APPROVED** [review]
Re-approving with lgtm.

Readded `PreserveUnknownFields` to the Metadata because odo devfile registries have devfiles with additional property in metadata and odo has already released with those registries. So this is needed to be backward compatible

File: pkg/devfile/header.go
Comment (davidfestal) [inline pkg/devfile/header.go:14]: @yangcao77 Though this PR is already merged, I just have a question about the reason for adding this line. It would allow adding arbitrary fields directly inside the `metadata` object without any way to retrieve them from the API. Isn't it precisely the situation we fixed by adding `attributes` one level down ?

Comment (yangcao77) [reply pkg/devfile/header.go:14]: Without this line,  the schema validation is going to fail when there are additional properties in metadata. 

The devfile registries for Odo 2.0 have devfiles ( [example](https://github.com/odo-devfiles/registry/blob/0b0e73401e462324a8db32b60c625f56ad639391/devfiles/nodejs/devfile.yaml#L5-L6) ) contain additional properties in metadata.  Without this line, using latest Odo (integrated with latest parser) with a created project would cause failure.  Since Odo 2.0 has already released with those registries,  devfile 2.0 changes need to be backward compatible.

Comment (davidfestal) [reply pkg/devfile/header.go:14]: OK, make perfectly sense. Thanks for the explanation !

**Pull request merged**
